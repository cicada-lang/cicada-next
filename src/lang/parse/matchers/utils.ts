import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
