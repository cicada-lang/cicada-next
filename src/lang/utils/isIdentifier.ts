const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
