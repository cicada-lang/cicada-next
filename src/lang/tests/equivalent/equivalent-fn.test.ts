import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("equivalent Fn", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent forall (T: Type, x: T) T {
    id
  = (T, x) => x
  = (T, x) => id(T, x)
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Fn -- alpha equivalent", async () => {
  const output = await runCode(`

compute equivalent forall (A: Type, B: Type) Type {
    (A, B) => A
  = (B, A) => B
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent FnAnnotated", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent forall (T: Type, x: T) T {
    id
  = (T, x) => x
  = (T, x) => id(T, x)
  = (T: Type, x: T) => x
  = (T: Type, x: T) => id(T, x)
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Fn -- fail", async () => {
  await expectCodeToFail(`

compute equivalent forall (A: Type, B: Type) Type {
    (A, B) => A
  = (A, B) => B
}

`)
})
