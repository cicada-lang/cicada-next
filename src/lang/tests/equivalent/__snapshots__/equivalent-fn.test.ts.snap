// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equivalent Fn > 
    "refl(
      implicit (T11: Type, T11) -> T11,
      implicit (T11, x11) => x11,
    ): Equal((T2: Type, T2) -> T2, (T2, x2) => x2, (T2, x2) => x2)"
   1`] = `
"refl(
  implicit (T11: Type, T11) -> T11,
  implicit (T11, x11) => x11,
): Equal((T2: Type, T2) -> T2, (T2, x2) => x2, (T2, x2) => x2)"
`;

exports[`equivalent Fn -- alpha equivalent > "refl(implicit (Type, Type) -> Type, implicit (A, B) => A): Equal((Type, Type) -> Type, (A, B) => A, (A, B) => A)" 1`] = `"refl(implicit (Type, Type) -> Type, implicit (A, B) => A): Equal((Type, Type) -> Type, (A, B) => A, (A, B) => A)"`;

exports[`equivalent Fn -- alpha equivalent: "refl(implicit (Type, Type) -> Type, implicit (A, B) => A): Equal((Type, Type) -> Type, (A, B) => A, (A, B) => A)" 1`] = `"refl(implicit (Type, Type) -> Type, implicit (A, B) => A): Equal((Type, Type) -> Type, (A, B) => A, (A, B) => A)"`;

exports[`equivalent Fn: 
    "refl(
      implicit (T11: Type, T11) -> T11,
      implicit (T11, x11) => x11,
    ): Equal((T2: Type, T2) -> T2, (T2, x2) => x2, (T2, x2) => x2)"
   1`] = `
"refl(
  implicit (T11: Type, T11) -> T11,
  implicit (T11, x11) => x11,
): Equal((T2: Type, T2) -> T2, (T2, x2) => x2, (T2, x2) => x2)"
`;

exports[`equivalent FnAnnotated > 
    "refl(
      implicit (T81: Type, T81) -> T81,
      implicit (T81, x81) => x81,
    ): Equal(
      (T11: Type, T11) -> T11,
      (T11, x11) => x11,
      (T11, x11) => x11,
    )"
   1`] = `
"refl(
  implicit (T81: Type, T81) -> T81,
  implicit (T81, x81) => x81,
): Equal(
  (T11: Type, T11) -> T11,
  (T11, x11) => x11,
  (T11, x11) => x11,
)"
`;

exports[`equivalent FnAnnotated: 
    "refl(
      implicit (T81: Type, T81) -> T81,
      implicit (T81, x81) => x81,
    ): Equal(
      (T11: Type, T11) -> T11,
      (T11, x11) => x11,
      (T11, x11) => x11,
    )"
   1`] = `
"refl(
  implicit (T81: Type, T81) -> T81,
  implicit (T81, x81) => x81,
): Equal(
  (T11: Type, T11) -> T11,
  (T11, x11) => x11,
  (T11, x11) => x11,
)"
`;

exports[`equivalent Fn 1`] = `
"refl(
  implicit (T11: Type, T11) -> T11,
  implicit (T11, x11) => x11,
): Equal((T2: Type, T2) -> T2, (T2, x2) => x2, (T2, x2) => x2)"
`;

exports[`equivalent Fn -- alpha equivalent 1`] = `"refl(implicit (Type, Type) -> Type, implicit (A, B) => A): Equal((Type, Type) -> Type, (A, B) => A, (A, B) => A)"`;

exports[`equivalent FnAnnotated 1`] = `
"refl(
  implicit (T81: Type, T81) -> T81,
  implicit (T81, x81) => x81,
): Equal(
  (T11: Type, T11) -> T11,
  (T11, x11) => x11,
  (T11, x11) => x11,
)"
`;
